File: code/library/src/main/java/org/fbme/lib/iec61499/stringify/STPrinter.java
package org.fbme.lib.iec61499.stringify;


import org.fbme.lib.st.expressions.*;
import org.fbme.lib.st.statements.*;

import java.util.List;

public class STPrinter {

    private String myIndent = "";
    private final StringBuilder myBuilder = new StringBuilder();

    private void append(String value) {
        myBuilder.append(value);
    }

    private void appendIndent() {
        myBuilder.append(myIndent);
    }

    private void appendNewLine() {
        myBuilder.append("\n");
    }

    private void withIndent(Runnable runnable) {
        String oldIndent = myIndent;
        myIndent += "  ";
        try {
            runnable.run();
        } finally {
            myIndent = oldIndent;
        }
    }

    private void appendExpression(Expression expression) {
        if (expression instanceof BinaryExpression) {
            var binaryExpression = (BinaryExpression) expression;
            appendExpression(binaryExpression.getLeftExpression());
            append(" ");
            append(binaryExpression.getOperation().getAlias());
            append(" ");
            appendExpression(binaryExpression.getRightExpression());
            return;
        }
        if (expression instanceof UnaryExpression) {
            var unaryExpression = (UnaryExpression) expression;
            var operation = unaryExpression.getOperation();
            append(operation.getAlias());
            append(operation.isSpaced() ? " " : "");
            appendExpression(unaryExpression.getInnerExpression());
            return;
        }
        if (expression instanceof FunctionCall) {
            var functionCall = (FunctionCall) expression;
            append(functionCall.getFunctionName());
            append("(");
            List<Expression> actualParameters = functionCall.getActualParameters();
            Expression last = actualParameters.get(actualParameters.size() - 1);
            for (Expression actualParameter : actualParameters) {
                appendExpression(actualParameter);
                if (actualParameter != last) {
                    append(", ");
                }
            }
            append(")");
            return;
        }
        if (expression instanceof ParenthesisExpression) {
            var parenthesisExpression = (ParenthesisExpression) expression;
            append("(");
            appendExpression(parenthesisExpression.getInnerExpression());
            append(")");
            return;
        }
        if (expression instanceof Variable) {
            appendVariable((Variable) expression);
            return;
        }
        if (expression instanceof Literal) {
            appendLiteral((Literal<?>) expression);
            return;
        }
        throw new IllegalArgumentException("Unrecognized expression " + expression);
    }

    private void appendVariable(Variable variable) {
        if (variable instanceof ArrayVariable) {
            var arrayVariable = (ArrayVariable) variable;
            appendVariable(arrayVariable.getSubscribedVariable());
            append("[");
            List<Expression> subscripts = arrayVariable.getSubscripts();
            Expression last = subscripts.get(subscripts.size() - 1);
            for (Expression subscript : subscripts) {
                appendExpression(subscript);
                if (subscript != last) {
                    append(", ");
                }
            }
            append("]");
            return;
        }
        if (variable instanceof VariableReference) {
            append(((VariableReference) variable).getReference().getPresentation());
            return;
        }
        throw new IllegalArgumentException("Unrecognized variable " + variable);
    }


    private void appendLiteral(Literal<?> literal) {
        var value = literal.getValue();
        switch (literal.getKind()) {
            case BINARY_BOOL:
                append(value == Boolean.TRUE ? "BOOL#1" : "BOOL#0");
                break;
            case BINARY_INT:
                append("2#" + Integer.toString((int) value, 2));
                break;
            case BOOL:
                append(value == Boolean.TRUE ? "TRUE" : "FALSE");
                break;
            case DEC_INT:
                append(Integer.toString((int) value));
                break;
            case HEX_INT:
                append("16#" + Integer.toString((int) value, 16));
                break;
            case OCT_INT:
                append("8#" + Integer.toString((int) value, 8));
                break;
            case STRING:
                append("'" + value + "'");
                break;
            case WSTRING:
                append("\"" + value + "\"");
                break;
            case TIME:
                append("T#" + value);
                break;
            default:
                throw new IllegalArgumentException("Unrecognized literal " + literal);
        }
    }

    private void appendIndentStatementList(List<Statement> statements) {
        withIndent(() -> {
            appendStatementList(statements);
        });
    }

    private void appendStatementList(List<Statement> statements) {
        for (Statement statement : statements) {
            appendStatement(statement);
        }
    }

    private void appendStatement(Statement statement) {
        appendIndent();
        try {
            if (statement instanceof AssignmentStatement) {
                var assignmentStatement = (AssignmentStatement) statement;
                appendVariable(assignmentStatement.getVariable());
                append(" := ");
                appendExpression(assignmentStatement.getExpression());
                return;
            }
            if (statement instanceof CaseStatement) {
                var caseStatement = (CaseStatement) statement;
                append("CASE ");
                appendExpression(caseStatement.getExpression());
                append(" OF ");
                appendNewLine();
                withIndent(() -> {
                    for (CaseElement element : caseStatement.getCases()) {
                        appendIndent();
                        appendLiteral(element.getLiteral());
                        append(" ->");
                        appendNewLine();
                        withIndent(() -> {
                            appendStatementList(element.getStatements());
                        });
                    }
                    List<Statement> elseCase = caseStatement.getElseCase();
                    if (elseCase != null) {
                        append("ELSE ->");
                        appendNewLine();
                        withIndent(() -> {
                            appendStatementList(elseCase);
                        });
                    }
                });
                appendIndent();
                append("END_CASE");
                return;
            }
            if (statement instanceof EmptyStatement) {
                return;
            }
            if (statement instanceof ExitStatement) {
                append("EXIT");
                return;
            }
            if (statement instanceof ForStatement) {
                var forStatement = (ForStatement) statement;
                ControlVariableDeclaration controlVariable = forStatement.getControlVariable();
                append("FOR ");
                append(controlVariable.getName());
                append(" := ");
                appendExpression(controlVariable.getBeginExpression());
                append(" TO ");
                appendExpression(controlVariable.getEndExpression());
                Expression stepExpression = controlVariable.getStepExpression();
                if (stepExpression != null) {
                    append(" BY ");
                    appendExpression(stepExpression);
                }
                append(" DO");
                appendNewLine();
                withIndent(() -> {
                    appendStatementList(forStatement.getStatements());
                });
                appendIndent();
                append("END_FOR");
            }
            if (statement instanceof IfStatement) {
                var ifStatement = (IfStatement) statement;
                append("IF ");
                appendExpression(ifStatement.getCondition());
                append(" THEN");
                appendNewLine();
                appendIndentStatementList(ifStatement.getThenClause());
                for (ElseIfClause elseIfClause : ifStatement.getElseIfClauses()) {
                    appendIndent();
                    append("ELSIF ");
                    appendExpression(elseIfClause.getCondition());
                    append(" THEN");
                    appendNewLine();
                    appendIndentStatementList(elseIfClause.getBody());
                }
                List<Statement> elseClause = ifStatement.getElseClause();
                if (elseClause != null) {
                    appendIndent();
                    append("ELSE");
                    appendNewLine();
                    appendIndentStatementList(elseClause);
                }
                appendIndent();
                append("END_IF");
            }
            if (statement instanceof RepeatStatement) {
                var repeatStatement = (RepeatStatement) statement;
                append("REPEAT");
                appendNewLine();
                withIndent(() -> {
                    appendStatementList(repeatStatement.getBody());
                });
                appendIndent();
                append("UNITL ");
                appendExpression(repeatStatement.getCondition());
                append(" END_REPEAT");
            }
            if (statement instanceof ReturnStatement) {
                append("RETURN");
            }
            if (statement instanceof WhileStatement) {
                var whileStatement = (WhileStatement) statement;
                append("WHILE ");
                appendExpression(whileStatement.getCondition());
                append(" DO");
                appendNewLine();
                withIndent(() -> {
                    appendStatementList(whileStatement.getBody());
                });
                appendIndent();
                append(" END_WHILE");
            }
        } finally {
            append(";");
            appendNewLine();
        }
    }


    public static String printStatementList(List<Statement> statements) {
        STPrinter printer = new STPrinter();
        printer.appendStatementList(statements);
        return printer.toString();
    }

    public static String printExpression(Expression expression) {
        STPrinter printer = new STPrinter();
        printer.appendExpression(expression);
        return printer.toString();
    }

    public static String printLiteral(Literal<?> literal) {
        STPrinter printer = new STPrinter();
        printer.appendLiteral(literal);
        return printer.toString();
    }

    @Override
    public String toString() {
        return myBuilder.toString();
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/stringify/STPrinter.kt
package org.fbme.lib.iec61499.stringify

import org.fbme.lib.st.expressions.*
import org.fbme.lib.st.statements.*
import kotlin.IllegalArgumentException
import kotlin.Int
import kotlin.String

class STPrinter {
    private var myIndent = ""
    private val myBuilder = StringBuilder()
    private fun append(value: String?) {
        myBuilder.append(value)
    }

    private fun appendIndent() {
        myBuilder.append(myIndent)
    }

    private fun appendNewLine() {
        myBuilder.append("\n")
    }

    private fun withIndent(runnable: Runnable) {
        val oldIndent = myIndent
        myIndent += "  "
        try {
            runnable.run()
        } finally {
            myIndent = oldIndent
        }
    }

    private fun appendExpression(expression: Expression?) {
        expression ?: return

        when (expression) {
            is BinaryExpression -> {
                appendExpression(expression.leftExpression)
                append(" ")
                append(expression.operation.alias)
                append(" ")
                appendExpression(expression.rightExpression)
            }

            is UnaryExpression -> {
                val operation = expression.operation
                append(operation.alias)
                append(if (operation.isSpaced) " " else "")
                appendExpression(expression.getInnerExpression())
            }

            is FunctionCall -> {
                append(expression.functionName)
                append("(")
                val actualParameters = expression.actualParameters
                val last = actualParameters[actualParameters.size - 1]
                for (actualParameter in actualParameters) {
                    appendExpression(actualParameter)
                    if (actualParameter !== last) {
                        append(", ")
                    }
                }
                append(")")
            }

            is ParenthesisExpression -> {
                append("(")
                appendExpression(expression.innerExpression)
                append(")")
            }

            is Variable -> {
                appendVariable(expression as Variable?)
            }

            is Literal<*> -> {
                appendLiteral(expression)
            }
        }
    }

    private fun appendVariable(variable: Variable?) {
        if (variable is ArrayVariable) {
            appendVariable(variable.subscribedVariable)
            append("[")
            val subscripts = variable.subscripts
            val last = subscripts[subscripts.size - 1]
            for (subscript in subscripts) {
                appendExpression(subscript)
                if (subscript !== last) {
                    append(", ")
                }
            }
            append("]")
            return
        }
        if (variable is VariableReference) {
            append(variable.reference.presentation)
            return
        }
        throw IllegalArgumentException("Unrecognized variable $variable")
    }

    private fun appendLiteral(literal: Literal<*>) {
        val value = literal.value
        when (literal.kind) {
            LiteralKind.BINARY_BOOL -> append(if (value == true) "BOOL#1" else "BOOL#0")
            LiteralKind.BINARY_INT -> append("2#" + (value as Int).toString(2))
            LiteralKind.BOOL -> append(if (value == true) "TRUE" else "FALSE")
            LiteralKind.DEC_INT -> append((value as Int).toString())
            LiteralKind.HEX_INT -> append("16#" + (value as Int).toString(16))
            LiteralKind.OCT_INT -> append("8#" + (value as Int).toString(8))
            LiteralKind.STRING -> append("'$value'")
            LiteralKind.WSTRING -> append("\"" + value + "\"")
            LiteralKind.TIME -> append("T#$value")
            LiteralKind.REAL -> append(value.toString())
            else -> throw IllegalArgumentException("Unrecognized literal $literal")
        }
    }

    private fun appendIndentStatementList(statements: List<Statement>) {
        withIndent { appendStatementList(statements) }
    }

    private fun appendStatementList(statements: List<Statement>) {
        for (statement in statements) {
            appendStatement(statement)
        }
    }

    private fun appendStatement(statement: Statement) {
        appendIndent()
        when (statement) {
            is AssignmentStatement -> {
                appendVariable(statement.variable)
                append(" := ")
                appendExpression(statement.expression)
            }

            is CaseStatement -> {
                append("CASE ")
                appendExpression(statement.expression)
                append(" OF ")
                appendNewLine()
                withIndent {
                    for (element in statement.cases) {
                        appendIndent()
                        appendLiteral(element.literal as Literal<*>)
                        append(" ->")
                        appendNewLine()
                        withIndent { appendStatementList(element.statements) }
                    }
                    val elseCase = statement.elseCase
                    if (elseCase != null) {
                        append("ELSE ->")
                        appendNewLine()
                        withIndent { appendStatementList(elseCase) }
                    }
                }
                appendIndent()
                append("END_CASE")
            }

            is EmptyStatement -> {
                return
            }

            is ExitStatement -> {
                append("EXIT")
            }

            is ForStatement -> {
                val controlVariable = statement.controlVariable
                append("FOR ")
                append(controlVariable.name)
                append(" := ")
                appendExpression(controlVariable.beginExpression)
                append(" TO ")
                appendExpression(controlVariable.endExpression)
                val stepExpression = controlVariable.stepExpression
                if (stepExpression != null) {
                    append(" BY ")
                    appendExpression(stepExpression)
                }
                append(" DO")
                appendNewLine()
                withIndent { appendStatementList(statement.statements) }
                appendIndent()
                append("END_FOR")
            }

            is IfStatement -> {
                append("IF ")
                appendExpression(statement.condition)
                append(" THEN")
                appendNewLine()
                appendIndentStatementList(statement.thenClause)
                for (elseIfClause in statement.elseIfClauses) {
                    appendIndent()
                    append("ELSIF ")
                    appendExpression(elseIfClause.condition)
                    append(" THEN")
                    appendNewLine()
                    appendIndentStatementList(elseIfClause.body)
                }
                val elseClause = statement.elseClause
                if (elseClause != null) {
                    appendIndent()
                    append("ELSE")
                    appendNewLine()
                    appendIndentStatementList(elseClause)
                }
                appendIndent()
                append("END_IF")
            }

            is RepeatStatement -> {
                append("REPEAT")
                appendNewLine()
                withIndent { appendStatementList(statement.body) }
                appendIndent()
                append("UNITL ")
                appendExpression(statement.condition)
                append(" END_REPEAT")
            }

            is ReturnStatement -> {
                append("RETURN")
            }

            is WhileStatement -> {
                append("WHILE ")
                appendExpression(statement.condition)
                append(" DO")
                appendNewLine()
                withIndent { appendStatementList(statement.body) }
                appendIndent()
                append(" END_WHILE")
            }
        }
        append(";")
        appendNewLine()
    }

    override fun toString(): String {
        return myBuilder.toString()
    }

    companion object {
        @JvmStatic
        fun printStatementList(statements: List<Statement>): String {
            val printer = STPrinter()
            printer.appendStatementList(statements)
            return printer.toString()
        }

        @JvmStatic
        fun printExpression(expression: Expression): String {
            val printer = STPrinter()
            printer.appendExpression(expression)
            return printer.toString()
        }

        @JvmStatic
        fun printLiteral(literal: Literal<*>): String {
            val printer = STPrinter()
            printer.appendLiteral(literal)
            return printer.toString()
        }
    }
}
